File: src/services/tauri/cc.ts
import { invokeTauri, isDesktopTauri, postNoContent, postJson, getJson } from './shared';

export async function ccNewSession(options: Record<string, unknown>, initialMessage: string) {
  if (isDesktopTauri()) {
    return await invokeTauri<string>('cc_new_session', { options, initialMessage });
  }
  return await postJson<string>('/api/cc/new-session', { options });
}

export async function ccSendMessage(sessionId: string, message: string) {
  if (isDesktopTauri()) {
    await invokeTauri('cc_send_message', { sessionId, message });
    return;
  }
  await postNoContent('/api/cc/send-message', {
    session_id: sessionId,
    message,
  });
}

export async function ccInterrupt(sessionId: string) {
  if (isDesktopTauri()) {
    await invokeTauri('cc_interrupt', { sessionId });
    return;
  }
  await postNoContent('/api/cc/interrupt', { session_id: sessionId });
}

export async function ccListSessions() {
  if (isDesktopTauri()) {
    return await invokeTauri<string[]>('cc_list_sessions');
  }
  return await getJson<string[]>('/api/cc/list-sessions');
}

export async function ccResumeSession(sessionId: string, options: Record<string, unknown>) {
  if (isDesktopTauri()) {
    await invokeTauri('cc_resume_session', { sessionId, options });
    return;
  }
  await postNoContent('/api/cc/resume-session', { session_id: sessionId, options });
}

export async function ccGetInstalledSkills() {
  if (isDesktopTauri()) {
    return await invokeTauri<string[]>('cc_get_installed_skills');
  }
  return await getJson<string[]>('/api/cc/installed-skills');
}

export async function ccGetSlashCommands(cwd?: string) {
  if (isDesktopTauri()) {
    return await invokeTauri<string[]>('cc_get_slash_commands', { cwd: cwd ?? null });
  }
  const qs = cwd ? `?cwd=${encodeURIComponent(cwd)}` : '';
  return await getJson<string[]>(`/api/cc/slash-commands${qs}`);
}

export async function ccGetProjects() {
  if (isDesktopTauri()) {
    return await invokeTauri<string[]>('cc_get_projects');
  }
  return await getJson<string[]>('/api/cc/projects');
}

export async function ccGetSessions<T = unknown>() {
  if (isDesktopTauri()) {
    return await invokeTauri<T>('cc_get_sessions');
  }
  return await getJson<T>('/api/cc/sessions');
}

export async function ccGetSettings<T = unknown>() {
  if (isDesktopTauri()) {
    return await invokeTauri<T>('cc_get_settings');
  }
  return await getJson<T>('/api/cc/settings');
}

export async function ccUpdateSettings(settings: unknown) {
  if (isDesktopTauri()) {
    await invokeTauri('cc_update_settings', { settings });
    return;
  }
  await postNoContent('/api/cc/settings', { settings });
}

export async function ccMcpAdd(request: unknown, workingDir: string) {
  if (isDesktopTauri()) {
    await invokeTauri('cc_mcp_add', { request, workingDir });
    return;
  }
  await postJson('/api/cc/mcp/add', { request, working_dir: workingDir });
}

export async function ccMcpList<T = unknown>(workingDir: string) {
  if (isDesktopTauri()) {
    return await invokeTauri<T>('cc_mcp_list', { workingDir });
  }
  return await postJson<T>('/api/cc/mcp/list', { working_dir: workingDir });
}

export async function ccMcpGet<T = unknown>(name: string, workingDir: string) {
  if (isDesktopTauri()) {
    return await invokeTauri<T>('cc_mcp_get', { name, workingDir });
  }
  return await postJson<T>('/api/cc/mcp/get', { name, working_dir: workingDir });
}

export async function ccListProjects() {
  if (isDesktopTauri()) {
    return await invokeTauri<string[]>('cc_list_projects');
  }
  return await getJson<string[]>('/api/cc/mcp/projects');
}

export async function ccMcpRemove(name: string, workingDir: string, scope = 'local') {
  if (isDesktopTauri()) {
    await invokeTauri('cc_mcp_remove', { name, workingDir, scope });
    return;
  }
  await postJson('/api/cc/mcp/remove', { name, working_dir: workingDir, scope });
}

export async function ccMcpEnable(name: string, workingDir: string) {
  if (isDesktopTauri()) {
    await invokeTauri('cc_mcp_enable', { name, workingDir });
    return;
  }
  await postJson('/api/cc/mcp/enable', { name, working_dir: workingDir });
}

export async function ccMcpDisable(name: string, workingDir: string) {
  if (isDesktopTauri()) {
    await invokeTauri('cc_mcp_disable', { name, workingDir });
    return;
  }
  await postJson('/api/cc/mcp/disable', { name, working_dir: workingDir });
}

export async function ccDeleteSession(sessionId: string): Promise<void> {
  if (isDesktopTauri()) {
    await invokeTauri('cc_delete_session', { sessionId });
    return;
  }
  await postNoContent('/api/cc/delete-session', { session_id: sessionId });
}

export async function ccGetSessionFilePath(sessionId: string): Promise<string | null> {
  if (isDesktopTauri()) {
    return await invokeTauri<string | null>('cc_get_session_file_path', { sessionId });
  }
  return await postJson<string | null>('/api/cc/session-file-path', { session_id: sessionId });
}

export async function ccResolvePermission(requestId: string, decision: string): Promise<void> {
  if (isDesktopTauri()) {
    return invokeTauri('cc_resolve_permission', { requestId, decision });
  } else {
    const port = import.meta.env.VITE_WEB_PORT || 8094;
    const response = await fetch(`http://127.0.0.1:${port}/api/cc/resolve-permission`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ request_id: requestId, decision }),
    });

    if (!response.ok) {
      throw new Error(`Failed to resolve permission: ${response.statusText}`);
    }
  }
}

export async function ccSetPermissionMode(sessionId: string, mode: string) {
  if (isDesktopTauri()) {
    await invokeTauri('cc_set_permission_mode', { sessionId, mode });
    return;
  }
  await postNoContent('/api/cc/set-permission-mode', { session_id: sessionId, mode });
}

File: src-tauri/src/commands/cc.rs
use codexia_cc::mcp::{
    ClaudeCodeMcpServer, ClaudeCodeResponse, cc_list_projects as mcp_cc_list_projects,
    cc_mcp_add as mcp_cc_mcp_add, cc_mcp_disable as mcp_cc_mcp_disable,
    cc_mcp_enable as mcp_cc_mcp_enable, cc_mcp_get as mcp_cc_mcp_get,
    cc_mcp_list as mcp_cc_mcp_list, cc_mcp_remove as mcp_cc_mcp_remove,
};
use codexia_cc::services::{
    message_service, session_service, settings_service, skill_service,
};
use codexia_cc::state::CCState;
use codexia_cc::types::{AgentOptions, CCConnectParams};
use session_service::SessionListResult;
use tauri::State;

#[tauri::command]
pub async fn cc_connect(
    params: CCConnectParams,
    state: State<'_, CCState>,
) -> Result<(), String> {
    session_service::connect(params, &state).await
}

#[tauri::command]
pub async fn cc_send_message(
    session_id: String,
    message: String,
    image_paths: Option<Vec<String>>,
    state: State<'_, CCState>,
) -> Result<(), String> {
    let image_paths = image_paths.unwrap_or_default();
    let cc_state = state.inner().clone();
    let sid = session_id.clone();
    message_service::send_message(&session_id, &message, &image_paths, &state, move |msg| {
        if let Ok(mut payload) = serde_json::to_value(&msg) {
            if let Some(obj) = payload.as_object_mut() {
                // Always override session_id with the caller's sid so that resumed sessions
                // (which may produce a new SDK session_id) are still routed to the correct card.
                obj.insert("session_id".to_string(), serde_json::Value::String(sid.clone()));
            }
            cc_state.emit("cc-message", payload);
        }
    })
    .await
}

#[tauri::command]
pub async fn cc_disconnect(session_id: String, state: State<'_, CCState>) -> Result<(), String> {
    session_service::disconnect(&session_id, &state).await
}

#[tauri::command]
pub async fn cc_new_session(
    options: AgentOptions,
    state: State<'_, CCState>,
) -> Result<String, String> {
    session_service::new_session(options, &state).await
}

#[tauri::command]
pub async fn cc_interrupt(session_id: String, state: State<'_, CCState>) -> Result<(), String> {
    session_service::interrupt(&session_id, &state).await
}

#[tauri::command]
pub async fn cc_resume_session(
    session_id: String,
    options: AgentOptions,
    state: State<'_, CCState>,
) -> Result<(), String> {
    session_service::resume_session(session_id, options, &state).await
}

#[tauri::command]
pub fn cc_get_installed_skills() -> Result<Vec<String>, String> {
    skill_service::get_installed_skills()
}

#[tauri::command]
pub fn cc_get_slash_commands(cwd: Option<String>) -> Result<Vec<String>, String> {
    skill_service::get_slash_commands(cwd.as_deref())
}

#[tauri::command]
pub fn cc_get_settings() -> Result<serde_json::Value, String> {
    settings_service::get_settings()
}

#[tauri::command]
pub fn cc_update_settings(settings: serde_json::Value) -> Result<(), String> {
    settings_service::update_settings(settings)
}

#[tauri::command]
pub fn cc_list_sessions(
    directory: Option<String>,
    limit: Option<usize>,
    offset: Option<usize>,
    include_worktrees: Option<bool>,
) -> Result<SessionListResult, String> {
    session_service::list_sessions(
        directory.as_deref(),
        limit,
        offset.unwrap_or(0),
        include_worktrees.unwrap_or(true),
    )
}

#[tauri::command]
pub fn cc_delete_session(session_id: String) -> Result<(), String> {
    claude_agent_sdk_rs::session_mutations::delete_session(&session_id, None)
        .map_err(|e| e.to_string())?;
    codexia_cc::db::SessionCache::new()?.delete_session(&session_id)?;
    Ok(())
}

#[tauri::command]
pub fn cc_get_session_messages(
    session_id: String,
) -> Result<Vec<claude_agent_sdk_rs::types::sessions::SessionMessage>, String> {
    Ok(claude_agent_sdk_rs::sessions::get_session_messages(&session_id, None, None, 0))
}

#[tauri::command]
pub async fn cc_mcp_list(working_dir: String) -> Result<Vec<ClaudeCodeMcpServer>, String> {
    mcp_cc_mcp_list(working_dir).await
}

#[tauri::command]
pub async fn cc_mcp_get(name: String, working_dir: String) -> Result<ClaudeCodeMcpServer, String> {
    mcp_cc_mcp_get(name, working_dir).await
}

#[tauri::command]
pub async fn cc_mcp_add(
    request: ClaudeCodeMcpServer,
    working_dir: String,
) -> Result<ClaudeCodeResponse, String> {
    mcp_cc_mcp_add(request, working_dir).await
}

#[tauri::command]
pub async fn cc_mcp_remove(
    name: String,
    working_dir: String,
    scope: String,
) -> Result<ClaudeCodeResponse, String> {
    mcp_cc_mcp_remove(name, working_dir, scope).await
}

#[tauri::command]
pub async fn cc_list_projects() -> Result<Vec<String>, String> {
    mcp_cc_list_projects().await
}

#[tauri::command]
pub async fn cc_mcp_disable(
    name: String,
    working_dir: String,
) -> Result<ClaudeCodeResponse, String> {
    mcp_cc_mcp_disable(name, working_dir).await
}

#[tauri::command]
pub async fn cc_mcp_enable(
    name: String,
    working_dir: String,
) -> Result<ClaudeCodeResponse, String> {
    mcp_cc_mcp_enable(name, working_dir).await
}
#[tauri::command]
pub async fn cc_resolve_permission(
    request_id: String,
    decision: String,
    state: State<'_, CCState>,
) -> Result<(), String> {
    state.resolve_permission(&request_id, decision)
}

#[tauri::command]
pub async fn cc_set_permission_mode(
    session_id: String,
    mode: String,
    state: State<'_, CCState>,
) -> Result<(), String> {
    session_service::set_permission_mode(&session_id, &mode, &state).await
}
